Generic Payload Protocol (#28)
mavlink/rfcs: text/todo-generic-payload.md

# RFC: Generic Payload Protocol for MAVLink

- Start Date: 2025-09-19
- Contributors: Thomas Watters (Skydio)
- Related Issues: N/A

## Summary

This RFC proposes a standardized Generic Payload Protocol for MAVLink that provides a flexible interface for controlling and monitoring arbitrary payloads attached to UAVs. The protocol supports various data types, control modes, and telemetry streaming through a set of five new MAVLink messages.

## Motivation

> Current MAVLink lacks a standardized way to interface with generic payloads that don't fit into existing message categories. While MAVLink does provide some [payload services](https://mavlink.io/en/services/payload.html), these have significant limitations for generic payload integration:

Reviewer: Can you comment on what types of payloads you're thinking of, so that we have some context what this is for and how it would work?

Author: will do, furthermore ill show how https://mavlink.io/en/services/illuminator.html can be completely replaced with this service, although the UI wouldn't necessarily be as first class.

Reviewer 2: Good question from Julian.

A bit of an aside, but we should also think of this in the context of the ROS 2 custom payload stuff Beat is doing for PX4 https://github.com/PX4/PX4-Autopilot/pull/25492#issue-3358387517 . The more we can know about a payload the better.

### Limitations of Existing MAVLink Payload Interface

1. **Payload-Specific Design**: Current commands are tailored to specific payload types (grippers, winches, illuminators) rather than providing a generic framework. The documentation explicitly recommends "use the commands for known payload types where possible as they are more intuitive for users."

2. **Limited Generic Commands**: The few generic commands available (`MAV_CMD_DO_SET_ACTUATOR`, `MAV_CMD_DO_SET_SERVO`) are:

   - Rudimentary in functionality (basic PWM/position control only)

>    - Inconsistently implemented ("only implemented on PX4" for actuator commands)

Reviewer 2: This is not something your proposal will solve, or could equally well be solved for the other APIs.

Author: with regards to what? the PX4 comment or that the generic commands are limited?

I believe my message set does clearly solve the issue of abstracting commands into rudimentary "functions".

Reviewer 2: What I'm saying is that it is painful that various methods are implemented on some FCs and not others, but that could equally well be true of this proposal.

Reviewer 2: > I believe my message set does clearly solve the issue of abstracting commands into rudimentary "functions".

Yes, it does IMO.

The real question is how needed it is seen to be by the flight stack communities. We already have dedicated APIs for parachutes, cameras, winches, grippers (halfway), illuminators, and other payloads. I'm not sure how many more payloads there are and whether the broad community would value the generic solution vs more custom solutions.

I certainly see the value in this over he bare actuator API - that is deliberately sparse because we wanted more tailored integrations. I don't know how the communities will feel. Once we/I have a good understanding we can start getting more feedback.

#### Control Modes

- `GENERIC_PAYLOAD_CONTROL_MODE`: Single-value enumeration for selecting control mode
  - `LATCHING` (1): Function maintains state until changed
  - `MOMENTARY` (2): Function auto-deactivates after timeout
- `GENERIC_PAYLOAD_CONTROL_MODE_FLAGS`: Bitmask for supported control modes
  - `SUPPORTS_LATCHING` (1): Function supports latching mode
  - `SUPPORTS_MOMENTARY` (2): Function supports momentary mode

#### Value Types

Support for 6 data types covering most payload needs:

- 32-bit: INT32, UINT32, REAL32
- 64-bit: INT64, UINT64, REAL64

#### Function Types

- `LOGICAL`: Boolean controls (on/off, enable/disable)
- `CONTINUOUS`: Analog controls with min/max ranges (servo angles, speeds)
- `DISCRETE`: Digital controls with enumerated values (modes, frequencies)

### Message ID Allocation

Proposed message IDs in the 60000-60004 range (vendor-specific range) to avoid conflicts during development. Final standardization would require official MAVLink message ID allocation.

### Extension Fields

All messages include extension fields for future enhancements:

- Power consumption monitoring
- Temperature sensing
- Mass/inertia properties
- 64-bit value support

> ## Alternatives

Reviewer 2: Might be worth pushing this to the end of the doc to not break up the flow.

### Alternative 1: Extend Existing Payload Commands

Build upon the current [MAVLink payload services](https://mavlink.io/en/services/payload.html) by adding more generic commands and standardizing implementation.

**Analysis**: The existing `MAV_CMD_DO_SET_ACTUATOR` and `MAV_CMD_DO_SET_SERVO` commands could theoretically be extended with additional parameters for type information and metadata.

**Key Limitations**:

- Commands are stateless and don't provide discovery mechanisms
- Inconsistent implementation across flight stacks ("only implemented on PX4")
- Limited parameter space in command messages prevents rich metadata
- No built-in telemetry streaming capabilities
- Would require significant breaking changes to existing command semantics

### Alternative 2: Parameter-Only Approach

> Use MAVLink's parameter system (`PARAM_*` messages) to expose payload functions as parameters.

Reviewer 2: MAVLink explicitly doesn't mandate what parameters are used - they are internal. The only exception is some of the [standardized names in the camera definition format](https://mavlink.io/en/services/camera_def.html#common-parameters) which need to be known in order to construct UIs. 

So while this is possible, I'd like to avoid it "directly"

Component metadata provides a mechanism to abstract parameter names - i.e. you can standardise a reference to a parameter name, and the component itself can name the parameter as it likes.

Author: indeed, this was the RAS-A approach although I don't think anyone has actually made use of it. To be clear this was an alternative considered, but not what I'm proposing we do.

Reviewer 2: Yes, I get that :-). I'm just pedantic.

**Key Limitations**:

- Parameters are designed for configuration, not real-time control
- No standardized discovery of which parameters belong to payloads
- Limited data types (mostly float32)
- No support for momentary operations or complex control modes
- Parameter namespace pollution with payload-specific entries
- No built-in telemetry streaming for sensor data

### Alternative 3: Extend Component Information Service

Build upon the [MAVLink Component Information Service](https://mavlink.io/en/services/component_information.html) to include payload metadata and add control messages.

**Analysis**: The component information service provides sophisticated metadata discovery through JSON files and supports actuator configuration. This could potentially be extended to describe payload capabilities.

**Key Limitations**:

- Currently "work in progress" with limited adoption
- Designed for static configuration metadata, not dynamic control
- Requires external file hosting or on-device storage for metadata
- JSON parsing overhead for simple operations
- No real-time control or telemetry streaming mechanisms

> - Metadata assumed invariant after boot, limiting dynamic payloads

Reviewer: This is what comes to mind when I see this proposal: can we extend COMPONENT_INFORMATION for this, and also use COMPONENT_INFORMATION_BASIC for the case where we don't want to use the JSON metadata.

Author: the problem would still exist that we would need a message set defined for each possible device type. The goal of the proposed message set is to support as many payloads as possible and support discovery. We do admit that the UX would not be a first class experience as something designed specifically for a device type (like illuminator)

Reviewer 2: Note, https://mavlink.io/en/messages/common.html#COMPONENT_METADATA (not COMPONENT_INFORMATION).
I will have a look at this PR tomorrow.

> - Metadata assumed invariant after boot, limiting dynamic payloads

Reviewer 2: Component metadata is essentially no longer WIP, but it does have limited implementation (PX4 only). ArduPilot have shown interest mostly for the metadata translation, but there is no pull for them, so they haven't pushed in their experimental branches.

There is an untested proposal/mechanism to allow dynamic update of metadata. This would support the main use case - dynamic addition of the payload. 

What this would be good for is providing metadata for your payload. You would still want to send the data out via normal MAVLink messages. The inspiration here would be the EVENT interface - it uses numbers to represent events and explain their meaning and the metadata describes those events. In your case this would allow you to provide all the static information about your payload, such as its name, and the function names, in metadata and then just use a number to identify the payload.

Author: That is clever, but what we are hoping to support here is the ability to develop a payload that is autonomous in its whole. So use of hosted metadata, or vehicle aware metadata to communicate functions/telemetry might not work, unless we force the payload to host the metadata locally and transmit on connection, but i think the _STATUS messages below serve that purpose

Reviewer 2: I get that. Just exploring the options :-)

## Unresolved Questions

> 1. **Message ID Range**: Should we request official MAVLink message IDs or remain in vendor-specific range?

Reviewer 2: This would go into development.xml with official message ids in the common range IFF RFC is broadly accepted. If not accepted you could still implement this in your own dialect.

2. **Backward Compatibility**: How should payloads indicate protocol version support?
3. **Performance Limits**: What are reasonable limits for number of functions/telemetry channels per payload?
4. **Rate Control**: How should high-rate telemetry interact with existing MAVLink rate limiting mechanisms?
5. **Error Recovery**: Should we define standard error recovery procedures for failed commands?

> 

Reviewer 2: 6. Support for FC connected payloads. 
   The camera and gripper (and soon to be winch) can also be attached to an FC. This means the payload can't have a unique component ID, and hence can't be addressed directly. We solved this for camera by having a camera id in the setter commands that can be set in missions to allow addressing of separate cameras in missions (where you don't have control over the the target) and also emitting this in the identity message. The first 6 ids are reserved for FC-attached payloads.

Author: I believe this would have to be handled at a manager layer right? fwiw our UX/UI does not communicate with MAVLink so we plan to have a layer between to translate to/from the attachment, although the intent was to also put the payload on the mavlink network as well.

Reviewer 2: Yes, the FC would have to implement code to process incoming messages and broadcast that it supports generic payloads of the given types. Typically it would also process the id in missions and direct the command to the hardware for attached payloads, or reemit the mission item as a command with the id value used as the component id.

An FC wouldn't have to support this - the only restriction this process adds is that generic payloads could not have a component id in the range reserved for those attached to an FC (that's arbitrary - for camera we used 1-6.

Error flags for a generic payload. This is a bitmask enumeration.

| Value | Name                                       | Description                                                                                                              |
| ----- | ------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------ |
| 1     | GENERIC_PAYLOAD_ERROR_FLAGS_SOFTWARE_ERROR | There is an error with the generic payload's software.                                                                   |
| 2     | GENERIC_PAYLOAD_ERROR_FLAGS_HARDWARE_ERROR | There is an error with the generic payload's hardware.                                                                   |
| 4     | GENERIC_PAYLOAD_ERROR_FLAGS_OVERTEMP_FAULT | A specified limit has been reached by the generic payload's temperature sensor. Applies if payload monitors temperature. |
| 8     | GENERIC_PAYLOAD_ERROR_FLAGS_CUSTOM         | Generic payload custom failure. See custom error flag bitmask for details.                                               |

### Messages

#### GENERIC_PAYLOAD_STATUS (Message ID: 60000)

> Status and discovery message for a generic payload. This should be broadcast at a low rate (nominally 1 Hz) to announce the payload's presence and capabilities.

Reviewer 2: I would be tempted to suggest that this be split into static and dynamic info. Use the https://mavlink.io/en/services/gimbal_v2.html#discovery-of-gimbal-manager approach based on MAV_TYPE to discover the payloads.

Not entirely sure, because it mixes status and metadata. Also I'm still not sure on how functions and payloads map in the real world, so maybe this is sensible.

Author: i think that makes sense.

Reviewer 2: > pproach based on MAV_TYPE to discover the payloads.

Though of course this might not work because you don't know the MAV_TYPE of a generic payload necessarily. More thought required.

| Field Name             | Type     | Description                                                          |

> | ---------------------- | -------- | -------------------------------------------------------------------- |

Reviewer 2: As per https://github.com/mavlink/rfcs/pull/28/files#r2373809642 probably a payload id to allow FC-attached payloads. Maybe remove the name and provide that as a metadata. Essentially, anything static _could_ be metadata or in some other low rate message.

Author: i don't think a payload id would be needed at this level, i could be convinced to add a payload id just now sure how thats any different than a component id.

Author: I agree we should move the metadata/static data to another message that is on request, which is essentially everything except for the  uptime_ms, error_flags, and the "extension fields"

Reviewer 2: > i don't think a payload id would be needed at this level, i could be convinced to add a payload id just now sure how thats any different than a component id.

For a MAVLink component the payload id is the same thing as a component id. So why do you need it, I hear you say? 
- Firstly, this allows you to address a particular generic component in a mission. The mission item messages don't include a target component field like command protocol, so if you don't add the id then the mission executor can really only broadcast the mission item to all components, or send to anything that it can identify as the desired type.
- For FC connected components the component id will be the autopilot id. So in this case you would need to have an ID so it can be resolved. This is an irritation of the limitations of FCs - in no way good MAVLink design, but it is forced upon us in many cases.

| uptime_ms              | uint32_t | Time since payload boot (in milliseconds).                           |
| error_flags            | uint32_t | Current error flags (bitmask). See GENERIC_PAYLOAD_ERROR_FLAGS.      |
| custom_error_flags     | uint32_t | Bitmap for custom error flags specific to this payload type.         |
| num_functions          | uint16_t | Number of functions this payload provides.                           |
| num_telemetry_channels | uint16_t | Number of telemetry channels this payload provides.                  |
| name                   | char[32] | Name of the generic payload for UI display (NULL-terminated string). |

> **Extensions:**

Reviewer 2: What do you mean by extensions? Separate messages or mavlink 2 extensions?

Reviewer 2: Either way, these are interesting things to consider.

Author: the intent was to be used as mavlink 2 extensions, though that may have been me overthinking it, wanted to minimize the size of the message if they weren't used.

Reviewer 2: I think these look like good things to expose, particularly for removable payloads. They might very reasonable also be useful for cameras and all the other payloads we have. In particular if they allow the vehicle to know that it perhaps has to do some retuning or thinking about how it has to fly.

I suspect though there will be many other things people want. TBD whether they should be extensions. My initial reaction is not. The mass and torque arm would be in the static payload message.

| max_low       | uint8_t[4] | Lower 32 bits of maximum value (for CONTINUOUS/DISCRETE types).                  |
| control_modes | uint16_t   | Bitmask of supported control modes. See GENERIC_PAYLOAD_CONTROL_MODE_FLAGS.      |
| timeout_ms    | uint32_t   | Timeout in milliseconds for MOMENTARY mode. 0 means use default timeout (100ms). |
| name          | char[32]   | Name of the function for UI display. NULL-terminated string.                     |
| units         | char[16]   | Units for the value field (optional). NULL-terminated string.                    |

**Extensions:**

| Field Name | Type       | Description                                                 |
| ---------- | ---------- | ----------------------------------------------------------- |
| value_high | uint8_t[4] | Upper 32 bits of current value. Only used for 64-bit types. |
| min_high   | uint8_t[4] | Upper 32 bits of minimum value. Only used for 64-bit types. |
| max_high   | uint8_t[4] | Upper 32 bits of maximum value. Only used for 64-bit types. |

#### GENERIC_PAYLOAD_FUNCTION_CONTROL (Message ID: 60002)

Command to control a generic payload function.

> 

Reviewer 2: If we have a payload id we will need to have that in here too. Obviously you don't need it if we were considering purely mavlink payloads.

| Value | Name                                       | Description                                                                                                              |
| ----- | ------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------ |
| 1     | GENERIC_PAYLOAD_ERROR_FLAGS_SOFTWARE_ERROR | There is an error with the generic payload's software.                                                                   |
| 2     | GENERIC_PAYLOAD_ERROR_FLAGS_HARDWARE_ERROR | There is an error with the generic payload's hardware.                                                                   |
| 4     | GENERIC_PAYLOAD_ERROR_FLAGS_OVERTEMP_FAULT | A specified limit has been reached by the generic payload's temperature sensor. Applies if payload monitors temperature. |
| 8     | GENERIC_PAYLOAD_ERROR_FLAGS_CUSTOM         | Generic payload custom failure. See custom error flag bitmask for details.                                               |

### Messages

#### GENERIC_PAYLOAD_DESCRIPTION (Message ID: 59999)

> Message broadcasted at a low rate (nominally < 1 Hz or on request) to announce the payload's presence and capabilities.

Reviewer 2: This is all static information. It occurs to me that you could:
1. Have a protocol bit to indicate support for generic payloads, then request from all components that have this. That's reasonable because you can't infer support from the MAV_TYPE as it could be anything.
2. It might be that you could have an url to a more precise component metadata for the capabilities here too. That would allow extension of non-generic metadata for a UI. For example, you could have a UI for parameters to be updated like the camera definition appoach.

Author: 1. I think GENERIC_PAYLOAD_DESCRIPTION indeed was the protocol bit to indicate the presence of this microservice

2. I'll add the URI field but it should definitely be optional. Our use case would be the vehicle consuming the payload's microservice and does not have a connection to the internet to pull down any metadata. furthermore we believe everything should be available statically from the payload.

| Field Name             | Type        | Description                                                          |
| ---------------------- | ----------- | -------------------------------------------------------------------- |

> | payload_id             | uint8_t     | Component ID of the generic payload.                                 |

Reviewer 2: Specify index from 1..

| num_functions          | uint16_t    | Number of functions this payload provides.                           |
| num_telemetry_channels | uint16_t    | Number of telemetry channels this payload provides.                  |
| name                   | char[32]    | Name of the generic payload for UI display (NULL-terminated string). |
| mass                   | uint16_t    | Mass of the payload in grams. 0 if not available.                    |

> | torque_arm             | uint16_t[3] | Torque arm in Payload Frame (mm)                                     |

Reviewer 2: I don't know what this means. Can it ever be not-available or invalid?

Author: yeah this is nuanced one that may not be relevant, i think the idea was to describe the CoM relative to the vehicles CoM or its Coordinate frame

#### GENERIC_PAYLOAD_STATUS (Message ID: 60000)

Status for a generic payload. This should be broadcast at a low rate (nominally 1 Hz) to announce the payload's presence and capabilities.

| Field Name             | Type     | Description                                                                          |
| ---------------------- | -------- | ------------------------------------------------------------------------------------ |
| payload_id             | uint8_t  | Component ID of the generic payload.                                                 |
| uptime_ms              | uint32_t | Time since payload boot (in milliseconds).                                           |
| error_flags            | uint32_t | Current error flags (bitmask). See GENERIC_PAYLOAD_ERROR_FLAGS.                      |
| custom_error_flags     | uint32_t | Bitmap for custom error flags specific to this payload type.                         |

> | power_draw             | uint16_t | Power consumption of the payload in mW. 0 if not available.                          |

Reviewer 2: ```suggestion
| power_draw             | uint16_t | Power consumption of the payload in mW. UINT16_MAX if not available.                          |
```

Note layed out properly, but makes more sense.

| temperature            | uint16_t | Temperature of the payload (0.01 °C per unit, 0 = 0°C). UINT16_MAX if not available. |

#### GENERIC_PAYLOAD_FUNCTION_DESCRIPTION (Message ID: 60001)

> Generic payload function description, sent on request with MAV_CMD_REQUEST_MESSAGE or MAV_CMD_SET_MESSAGE_INTERVAL, with param2 set to payload_id, and param3 set to function index.

Reviewer 2: ```suggestion
Generic payload function description, sent on request with MAV_CMD_REQUEST_MESSAGE or MAV_CMD_SET_MESSAGE_INTERVAL, with param2 set to payload_id. Param 3 can be set to 0 to trigger emission of the message for all functions, or a function index to get information for that function.
```

| Field Name    | Type       | Description                                                                      |
| ------------- | ---------- | -------------------------------------------------------------------------------- |
| payload_id    | uint8_t    | Component ID of the generic payload.                                 |
| index         | uint16_t   | Index of this function on the generic payload.                                   |
| type          | uint8_t    | Type of function. See GENERIC_PAYLOAD_FUNCTION_TYPE.                             |
| value_type    | uint8_t    | Data type for value/min/max fields. See GENERIC_PAYLOAD_VALUE_TYPE.              |
| enabled       | uint8_t    | 0: Disabled, 1: Enabled                                                          |
| min_low       | uint8_t[4] | Lower 32 bits of minimum value (for CONTINUOUS/DISCRETE types). Little-endian.   |

> | max_low       | uint8_t[4] | Lower 32 bits of maximum value (for CONTINUOUS/DISCRETE types). Little-endian.   |

Reviewer 2: Presumably set to all 0 for these and min values for boolean type or an enum?

Author: i assumed bool min/max 0/1 and for enum it would just cover the range. i'm not sure it matters to much because the UX/UI could choose to display these in some way. Unfortunately with enums or bitmask there isn't a generic trivial way to describe the names of the values that was lightweight enough.

| name          | char[32]   | Name of the function for UI display. NULL-terminated string.                     |
| units         | char[16]   | Units for the value field (optional). NULL-terminated string.                    |
| min_high      | uint8_t[4] | Upper 32 bits of minimum value. Only used for 64-bit types. Little-endian        |
| max_high      | uint8_t[4] | Upper 32 bits of maximum value. Only used for 64-bit types. Little-endian        |

#### GENERIC_PAYLOAD_FUNCTION_STATUS (Message ID: 60001)

Generic payload function status.

| Field Name    | Type       | Description                                                                                     |
| ------------- | ---------- | ----------------------------------------------------------------------------------------------- |
| payload_id    | uint8_t    | Component ID of the generic payload.                                                            |
| index         | uint16_t   | Index of this function on the generic payload.                                                  |
| value_low     | uint8_t[4] | Lower 32 bits of current value. For 32-bit types, this is the entire value. Little-endian.      |

> | value_high    | uint8_t[4] | Upper 32 bits of current value. Only used for 64-bit types. Little-endian.                      |

Reviewer 2: Does this cover all the types? What should we do for a boolean type?

Author: send a value_low != 0 or value_low == 1

#### GENERIC_PAYLOAD_FUNCTION_CONTROL (Message ID: 60002)

> Command to control a generic payload function.

Reviewer 2: This makes sense rather than a command, given the fields. But it does have two limitations. This means the feature can't be used in missions. It also has no ACK, and no target device or or component for addressing - you would need to add fields for these.

Author: we could technically wrap this in a CMD if we came up with a COMMAND_VAR, or something that had a variable length array. but i don't think that is needed as part of this RFC

Reviewer 2: > but i don't think that is needed as part of this RFC

I appreciate you have an internal implementation that does need this, but IMO support in missions is likely to be a dealbreaker for acceptance into common. So is having an ACK for acceptance of the command. 
This maps quite easily conceptually as shown.

```
      <entry value="60000" name="MAV_CMD_GENERIC_PAYLOAD_FUNCTION" hasLocation="false" isDestination="false">
        <description>Command to control a generic payload function.</description>
        <param index="1" label="Payload id" minValue="1" maxValue="255">Component ID of the generic payload. Used in missions only.</param>
        <param index="2" label="Index" minValue="1">Index of this function on the generic payload.</param>
        <param index="3" label="Control mode" enum="GENERIC_PAYLOAD_CONTROL_MODE">Desired Control Mode.</param>
        <param index="4" label="Enable" enum="MAV_BOOL">1: Enable, 2: Disable.</param>
        <param index="6" label="Value_low" units = "ms" minValue="0">Lower 32 bits of value to set. For 32-bit types, this is the entire value. Little-endian.</param>
        <param index="7" label="Value_high" minValue="0">Upper 32 bits of value to set. Only used for 64-bit types. Little-endian.</param>
        <param index="8" label="Timeout" minValue="0">Timeout for momentary operation if control_mode is GENERIC_PAYLOAD_CONTROL_MODE_MOMENTARY.</param>
      </entry>
```

Author: I'm ok with adding a mav_cmd, but I think there is also a benefit to having a control message that can be streamed/sent more frequently than the command protocol allows, so I think both should be added/documented properly.

| Field Name   | Type       | Description                                                                                |
| ------------ | ---------- | ------------------------------------------------------------------------------------------ |
| payload_id   | uint8_t    | Component ID of the generic payload.                                                       |
| index        | uint16_t   | Index of this function on the generic payload.                                             |
| control_mode | uint8_t    | Desired Control Mode. See GENERIC_PAYLOAD_CONTROL_MODE.                                    |
| enable       | uint8_t    | 0: Disable, 1: Enable                                                                      |
| value_low    | uint8_t[4] | Lower 32 bits of value to set. For 32-bit types, this is the entire value. Little-endian.  |
| timeout_ms   | uint32_t   | Timeout for momentary operation if control_mode is GENERIC_PAYLOAD_CONTROL_MODE_MOMENTARY. |
| value_high   | uint8_t[4] | Upper 32 bits of value to set. Only used for 64-bit types. Little-endian.                  |

#### GENERIC_PAYLOAD_TELEMETRY_DESCRIPTION (Message ID: 60003)

> Generic payload telemetry channel description, sent on request with MAV_CMD_REQUEST_MESSAGE or MAV_CMD_SET_MESSAGE_INTERVAL, with param2 set to payload_id, and param3 set to function index.

Reviewer 2: Does this/should this match  GENERIC_PAYLOAD_FUNCTION_DESCRIPTION? If so, can't we use that for this?

Author: i think i like them separate just for consistency, ie the FUNCTION set of messages and the TELEMETRY set.

to push them into the same message we'd have to add a field to indicate function/telem, and then reuses enable/update_rate field

Reviewer 2: OK. This is likely to come up again. Flight stacks prefer fewer messages. Let's see.

- mode
- brightness
- strobe_period (min_strobe_period, max_strobe_period in function description)
- strobe_duty_cycle

### MAV_CMD_ILLUMINATOR_ON_OFF

**GENERIC_PAYLOAD_FUNCTION_DESCRIPTION**

On/Off:
- index = 0
- type = GENERIC_PAYLOAD_FUNCTION_TYPE_LOGICAL
- value_type = GENERIC_PAYLOAD_VALUE_TYPE_UINT32
- enabled = 1
- min_low = 0
- max_low = 1
- control_modes = GENERIC_PAYLOAD_SUPPORTS_LATCHING
- timeout_ms = 0
- name = "On/Off"
- units = ""

**GENERIC_PAYLOAD_FUNCTION_STATUS**

On:
- payload_id = 1
- index = 0
- value_low = union(uint32_t, uint8_t b = 1) -> b

Off:
- payload_id = 1
- index = 0
- value_low = union(uint32_t, uint8_t b = 0) -> b
 
**GENERIC_PAYLOAD_FUNCTION_CONTROL**

On:
- payload_id = 1
- index = 0
- control_mode = GENERIC_PAYLOAD_CONTROL_MODE_LATCHING
- enable = 1
- value_low = union(uint32_t, uint8_t b = 1) -> b

Off:
- payload_id = 1
- index = 0
- control_mode = GENERIC_PAYLOAD_CONTROL_MODE_LATCHING
- enable = 1
- value_low = union(uint32_t, uint8_t b = 0) -> b

### MAV_CMD_DO_ILLUMINATOR_CONFIGURE

#### Mode

**GENERIC_PAYLOAD_FUNCTION_DESCRIPTION**

- index = 1
- type = GENERIC_PAYLOAD_FUNCTION_TYPE_BITMASK

> - value_type = GENERIC_PAYLOAD_VALUE_TYPE_BITMASK8

Reviewer 2: ```suggestion
- value_type = GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_8
```

| 5     | GENERIC_PAYLOAD_VALUE_TYPE_REAL64 | 64-bit IEEE-754 floating point. |
| 6     | GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_8  | Bitmask of values.          |
| 7     | GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_16 | Bitmask of values.          |
| 8     | GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_32 | Bitmask of values.          |
| 9     | GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_64 | Bitmask of values.          |

#### GENERIC_PAYLOAD_FUNCTION_TYPE

Type of generic payload function.

| Value | Name                                     | Description                                                                |
| ----- | ---------------------------------------- | -------------------------------------------------------------------------- |
| 0     | GENERIC_PAYLOAD_FUNCTION_TYPE_LOGICAL    | Logical state control (true/false, on/off, high/low).                      |
| 1     | GENERIC_PAYLOAD_FUNCTION_TYPE_CONTINUOUS | Continuous value control (e.g., servo angle, motor speed). Uses min/max.   |
| 2     | GENERIC_PAYLOAD_FUNCTION_TYPE_DISCRETE   | Discrete value selection (e.g., mode number, PWM frequency). Uses min/max. |

> | 3     | GENERIC_PAYLOAD_FUNCTION_TYPE_BITMASK    | Bitmask value selection                                                    |

Reviewer 2: The bitmask and enum information obviously is trivially easy with metadata. The only way I can see to do this in messages is  to send a message for metadata about a function type. This needs strings for each bit or enum values. Given the possible size of strings, probably do this in groups of 8. It could be quite efficient - you can know how many there will be and you only need to request that many and once for each service.

The message would use a value_index  to allow for bits/values greater than 8. Strings still very limited.
Thoughts?

#### GENERIC_PAYLOAD_FUNCTION_VALUE_METADATA_DESCRIPTION (Message ID: TBD)

Generic payload for describing the values of functions that are of DISCRETE (enum) or bit type, sent on request with MAV_CMD_REQUEST_MESSAGE. The value index allows you to specify that this is the 0th, 1st, etc byte in the function, or for the 9th, 17th or so on enum value, and so on.
This does limit the descriptions. We might in theory allow some to be smaller and some larger, but it complicates things.

| Field Name    | Type       | Min | Description                                                |
| ------------- | ---------- | --- | -----------------------------------------------------------|
| payload_id    | uint8_t    | 1   | Component ID of the generic payload.                       |
| index         | uint16_t   | 1   | Index of this function on the generic payload.             |
| value_index   | uint8_t    | 1   | Index the first byte or value.                              |
| item 1        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 2        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 3        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 4        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 5        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 6        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 7        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |
| item 8        | char[31]   |     | Name of the item for UI display. NULL-terminated string.    |

Author: yeah this would be a very verbose way to give names to the bitmasks/enumerations, i think we can leave this to the metadata

Reviewer 2: Yeah, but once you go down requiring support for metadata you might as well say all of the description should be in metadata.

| payload_id             | uint8_t     | 1   | Component ID of the generic payload.                                                                              |
| num_functions          | uint16_t    | 0   | Number of functions this payload provides.                                                                        |
| num_telemetry_channels | uint16_t    | 0   | Number of telemetry channels this payload provides.                                                               |
| name                   | char[32]    |     | Name of the generic payload for UI display (NULL-terminated string).                                              |
| mass                   | uint16_t    |     | Mass of the payload in grams. 0 if not available.                                                                 |
| torque_arm             | uint16_t[3] |     | Vector3 describing the CoM offset from the vehicle's payload frame CoM (mm). Optional, 0s indicate not specified. |
| metadata_crc           | uint32_t    |     | CRC32 checksum of the metadata file. 0 if metadata_uri is not used.                                               |

> | metadata_uri           | char[100]   |     | URI to component metadata file (NULL-terminated string). All zeros if not used.                                   |

Reviewer 2: Perhaps  add "if specified, it should define all the information in all the metadata. " The metadata messages would therefore not be needed and should not be emitted" ?

We should define the JSON for this.

Author: i'm not sure we should stop emitting the protocol messages if the metadata_uri exists. it could be possible that a consumer of the protocol doesn't know how to consume metadata or can't

- mode
- brightness
- strobe_period (min_strobe_period, max_strobe_period in function description)
- strobe_duty_cycle

### MAV_CMD_ILLUMINATOR_ON_OFF

**GENERIC_PAYLOAD_FUNCTION_DESCRIPTION**

On/Off:
- index = 0
- type = GENERIC_PAYLOAD_FUNCTION_TYPE_LOGICAL
- value_type = GENERIC_PAYLOAD_VALUE_TYPE_UINT32
- enabled = 1
- min_low = 0
- max_low = 1
- control_modes = GENERIC_PAYLOAD_SUPPORTS_LATCHING
- timeout_ms = 0
- name = "On/Off"
- units = ""

**GENERIC_PAYLOAD_FUNCTION_STATUS**

On:
- payload_id = 1
- index = 0
- value_low = union(uint32_t, uint8_t b = 1) -> b

Off:
- payload_id = 1
- index = 0
- value_low = union(uint32_t, uint8_t b = 0) -> b
 
**GENERIC_PAYLOAD_FUNCTION_CONTROL**

On:
- payload_id = 1
- index = 0
- control_mode = GENERIC_PAYLOAD_CONTROL_MODE_LATCHING
- enable = 1
- value_low = union(uint32_t, uint8_t b = 1) -> b

Off:
- payload_id = 1
- index = 0
- control_mode = GENERIC_PAYLOAD_CONTROL_MODE_LATCHING
- enable = 1
- value_low = union(uint32_t, uint8_t b = 0) -> b

### MAV_CMD_DO_ILLUMINATOR_CONFIGURE

#### Mode

**GENERIC_PAYLOAD_FUNCTION_DESCRIPTION**

- index = 1
- type = GENERIC_PAYLOAD_FUNCTION_TYPE_BITMASK

> - value_type = GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_8

Reviewer 2: How would I provide metadata to represent a bitmask or the allowed values of an enum?

Reviewer 2: @thomas-watters-skydio Can you comment on this?

Author: I don't have a great solution for this. I would assume it would need to be communicated apriori in some way alongside the attachment or through a component metadata service. Open to ideas for it, this was only added to support further configuration functions currently seen in other micro services.

- Requires external file hosting or on-device storage for metadata
- JSON parsing overhead for simple operations
- No real-time control or telemetry streaming mechanisms
- Metadata assumed invariant after boot, limiting dynamic payloads

## Unresolved Questions

1. **Message ID Range**: Should we request official MAVLink message IDs or remain in vendor-specific range?
2. **Backward Compatibility**: How should payloads indicate protocol version support?
3. **Performance Limits**: What are reasonable limits for number of functions/telemetry channels per payload?
4. **Rate Control**: How should high-rate telemetry interact with existing MAVLink rate limiting mechanisms?
5. **Error Recovery**: Should we define standard error recovery procedures for failed commands?

## References

- [MAVLink Message Definition Standard](https://mavlink.io/en/guide/define_xml_element.html)
- [MAVLink Common Message Set](https://mavlink.io/en/messages/common.html)
- [Existing Payload Interfaces in MAVLink](https://mavlink.io/en/messages/common.html#camera-protocol)
- [MAVLink Extension Fields](https://mavlink.io/en/guide/message_definitions.html#extension_fields)

> - [MAVLink Message Rate Control](https://mavlink.io/en/messages/common.html#MAV_CMD_SET_MESSAGE_INTERVAL)

Reviewer 2: I started looking at how you would do this using component metadata for the [MAV_CMD_DO_WINCH ](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_WINCH) and associated telemetry. By no means complete, but started as below.

The command has a number of operations you can do, which are defined by a value set in actions. That would be one mapped function.

Have the MAV_CMD also allows you to set the line length and rate for that command.
My assumption is that you would map this as three commands, one each for setting the lenght and rate and the other for doing the action.
1. Does that make sense? (Its the same for when using the messages.
2. This isn't really doing an action - its just setting a value. Can you show me how you would do setting the line length?

```json
{
  "generic_device_protocol": [
    {
      "payload_id": 1,
      "name": "Winch",
      "vendor": "Generic",
      "model": "AutoWinch-1000",
      "serial_number": "WNCH0001",
      "software_version": "1.0.0",
      "hardware_version": "1.0",
      "mass": 1200,
      "torque_arm": { "x": 0, "y": 0, "z": 50 },
      "description": "Autonomous winch payload capable of deploying and retrieving tethered loads.",

      "functions": [
        {
          "id": 1,
          "name": "Action",
          "type": "GENERIC_PAYLOAD_FUNCTION_TYPE_DISCRETE",
          "enabled": 1,
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_UINT32",
          "min": 0,
          "max": 7,
          "control_modes": ["GENERIC_PAYLOAD_SUPPORTS_MOMENTARY"],
          "timeout_ms": 100,
          "units": "enum:WINCH_ACTIONS",
          "description": "Perform a winch action defined by WINCH_ACTIONS (e.g., RELAXED, RETRACT, HOLD, etc.)."
        },
        {
          "id": 2,
          "name": "Set Line Length",
          "type": "GENERIC_PAYLOAD_FUNCTION_TYPE_DISCRETE",
          "enabled": 1,
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "min": 0.0,
          "max": 50.0,
          "control_modes": ["GENERIC_PAYLOAD_SUPPORTS_LATCHING"],
          "timeout_ms": 0,
          "units": "m",
          "description": "Sets the target line length (negative not allowed)."
        },
        {
          "id": 3,
          "name": "Set Line Rate",
          "type": "GENERIC_PAYLOAD_FUNCTION_TYPE_DISCRETE",
          "enabled": 1,
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "min": -2.0,
          "max": 2.0,
          "control_modes": ["GENERIC_PAYLOAD_SUPPORTS_LATCHING"],
          "timeout_ms": 0,
          "units": "m/s",
          "description": "Sets the target cable rate (positive to release, negative to retract)."
        }
      ],

      "telemetry_channels": [
        {
          "channel_id": 1,
          "name": "Cable Length",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "update_rate": 5,
          "min": 0,
          "max": 50,
          "units": "m",
          "description": "Cable length deployed, derived from WHEEL_DISTANCE.distance."
        },
        {
          "channel_id": 2,
          "name": "Cable Speed",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "update_rate": 5,
          "min": -2,
          "max": 2,
          "units": "m/s",
          "description": "Winch cable rate of movement."
        },
        {
          "channel_id": 3,
          "name": "Tension",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "update_rate": 5,
          "min": 0,
          "max": 1000,
          "units": "N",
          "description": "Current measured cable tension from WINCH_STATUS.tension."
        },
        {
          "channel_id": 4,
          "name": "Voltage",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "update_rate": 1,
          "min": 0,
          "max": 30,
          "units": "V",
          "description": "Motor or system voltage from WINCH_STATUS.voltage."
        },
        {
          "channel_id": 5,
          "name": "Current",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "update_rate": 1,
          "min": 0,
          "max": 50,
          "units": "A",
          "description": "Motor current draw from WINCH_STATUS.current."
        },
        {
          "channel_id": 6,
          "name": "Temperature",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_REAL32",
          "update_rate": 1,
          "min": -40,
          "max": 125,
          "units": "°C",
          "description": "Motor or drum temperature from WINCH_STATUS.temperature."
        },
        {
          "channel_id": 7,
          "name": "Status Flags",
          "value_type": "GENERIC_PAYLOAD_VALUE_TYPE_BITMASK_32",
          "update_rate": 1,
          "min": 0,
          "max": 8192,
          "description": "Bitmask from WINCH_STATUS.status_flags indicating winch operational state.",
          "allowed_values": [
            { "value": 1, "name": "MAV_WINCH_STATUS_HEALTHY", "description": "Winch is healthy." },
            { "value": 2, "name": "MAV_WINCH_STATUS_FULLY_RETRACTED", "description": "Winch line is fully retracted." },
            { "value": 4, "name": "MAV_WINCH_STATUS_MOVING", "description": "Winch motor is moving." },
            { "value": 8, "name": "MAV_WINCH_STATUS_CLUTCH_ENGAGED", "description": "Clutch engaged, allowing motor to move freely." },
            { "value": 16, "name": "MAV_WINCH_STATUS_LOCKED", "description": "Winch is locked by the locking mechanism." },
            { "value": 32, "name": "MAV_WINCH_STATUS_DROPPING", "description": "Winch is gravity dropping payload." },
            { "value": 64, "name": "MAV_WINCH_STATUS_ARRESTING", "description": "Winch is arresting payload descent." },
            { "value": 128, "name": "MAV_WINCH_STATUS_GROUND_SENSE", "description": "Using torque to sense ground." },
            { "value": 256, "name": "MAV_WINCH_STATUS_RETRACTING", "description": "Returning to fully retracted position." },
            { "value": 512, "name": "MAV_WINCH_STATUS_REDELIVER", "description": "Redelivering payload (failover if tension threshold exceeded)." },
            { "value": 1024, "name": "MAV_WINCH_STATUS_ABANDON_LINE", "description": "Abandoning line/payload (failover from REDELIVER)." },
            { "value": 2048, "name": "MAV_WINCH_STATUS_LOCKING", "description": "Engaging the locking mechanism." },
            { "value": 4096, "name": "MAV_WINCH_STATUS_LOAD_LINE", "description": "Spooling on line." },
            { "value": 8192, "name": "MAV_WINCH_STATUS_LOAD_PAYLOAD", "description": "Loading a payload." }
          ]
        }
      ]
    }
  ]
}
```

Author: what you have makes sense, so i need to come up with a schema that allows you to communicate the generic_payload_protocol in its entirey?

> This isn't really doing an action - its just setting a value. Can you show me how you would do setting the line length?

I don't really follow, i think how you described it is correct, settings would have to map to a "function"

Reviewer 2: > so i need to come up with a schema that allows you to communicate the generic_payload_protocol in its entirey?

FWIW I think this protocol makes way more sense as component metadata except in one respect: ArduPilot do not yet implement the concept. 
In one sense though that doesn't matter; for this to work all that is really needed is that the component exposing the data is able to share the file via mavftp. If the data is static then it can be added to the firmware at build time. 
The recipient has to understand the data format, of course, but QGC already supports the protocol (MISSION planner does not, but this might force it to).

That said, the answer to your question is not that you have to do anything - but it would be useful. The reason I did this was to test the feasibility once I ran into the fact that you can't document flags or enum values using the current approach.

>> This isn't really doing an action - its just setting a value. Can you show me how you would do setting the line length?
> I don't really follow, i think how you described it is correct, settings would have to map to a "function"

Thanks. I just wanted to check that in order to emulate current winch functionality you would need to break this into three functions. If one of those functions is setting a parameter then you would also need to implement the setting parameter as a function, and presumably that function would be associated with the particular device instance.

This makes things a bit complicated for the user because it isn't obvious that deploying a winch is a matter of setting the line length, the action, the rate, This would be presented to them as a flat list of functions, which wouldn't necessarily be ordered.

W.r.t. the format above, I was just looking at the data at this point. But as a next stage we might look at formatting how the data is presented as part of the syntax.
For example, it might be possible to indicate that the three functions are grouped, and that they should be executed as such or rendered differently because of that. It might be possible to decide that some functions depend on other values, and so on.

The upshot is that I feel we really need implementations to test this in the UI to know what will work. It is all a bit abstract right now.
